scheds/rust/scx_p2dq/src/bpf/main.bpf.c

```diff
@@ -0,0 +1,1006 @@
+private(A) struct bpf_cpumask __kptr *big_cpumask;
+
+
+static __always_inline u64 dsq_time_slice(int dsq_id)
+{
+	if (dsq_id > nr_dsqs_per_llc || dsq_id < 0) {
+		scx_bpf_error("Invalid DSQ id %d", dsq_id);
+		return 0;
+	}
+	return dsq_time_slices[dsq_id];
+}
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+	__type(key, u32);
+	__type(value, struct cpu_ctx);
+	__uint(max_entries, 1);
+} cpu_ctxs SEC(".maps");
+
+static struct cpu_ctx *lookup_cpu_ctx(int cpu)
+{
+	struct cpu_ctx *cpuc;
+
+	if (cpu < 0)
+		cpuc = bpf_map_lookup_elem(&cpu_ctxs, &zero_u32);
+	else
+		cpuc = bpf_map_lookup_percpu_elem(&cpu_ctxs, &zero_u32, cpu);
+
+	if (!cpuc) {
+		scx_bpf_error("no cpu_ctx for cpu %d", cpu);
+		return NULL;
+	}
+
+	return cpuc;
+}
+
+static __always_inline u64 cpu_dsq_id(int dsq_index, struct cpu_ctx *cpuc) {
+	if (dsq_index < 0 || dsq_index > nr_dsqs_per_llc || dsq_index >= MAX_DSQS_PER_LLC) {
+		scx_bpf_error("invalid dsq index: %d", dsq_index);
+		return 0;
+	}
+	return *MEMBER_VPTR(cpuc->dsqs, [dsq_index]);
+}
+
+
+struct {
+	__uint(type, BPF_MAP_TYPE_ARRAY);
+	__type(key, u32);
+	__type(value, struct llc_ctx);
+	__uint(max_entries, MAX_LLCS);
+} llc_ctxs SEC(".maps");
+
+static struct llc_ctx *lookup_llc_ctx(u32 llc_id)
+{
+	struct llc_ctx *llcx;
+
+	llcx = bpf_map_lookup_elem(&llc_ctxs, &llc_id);
+	if (!llcx) {
+		scx_bpf_error("no llc_ctx for llc %u", llc_id);
+		return NULL;
+	}
+
+	return llcx;
+}
+
+
+struct {
+	__uint(type, BPF_MAP_TYPE_ARRAY);
+	__type(key, u32);
+	__type(value, struct node_ctx);
+	__uint(max_entries, MAX_NUMA_NODES);
+	__uint(map_flags, 0);
+} node_ctxs SEC(".maps");
+
+static struct node_ctx *lookup_node_ctx(u32 node_id)
+{
+	struct node_ctx *nodec;
+
+	nodec = bpf_map_lookup_elem(&node_ctxs, &node_id);
+	if (!nodec) {
+		scx_bpf_error("no node_ctx for node %u", node_id);
+		return NULL;
+	}
+
+	return nodec;
+}
+
+
+struct {
+	__uint(type, BPF_MAP_TYPE_TASK_STORAGE);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+	__type(key, int);
+	__type(value, struct task_ctx);
+} task_ctxs SEC(".maps");
+
+static struct task_ctx *lookup_task_ctx_may_fail(struct task_struct *p)
+{
+	return bpf_task_storage_get(&task_ctxs, p, 0, 0);
+}
+
+static struct task_ctx *lookup_task_ctx(struct task_struct *p)
+{
+	struct task_ctx *taskc = lookup_task_ctx_may_fail(p);
+
+	if (!taskc)
+		scx_bpf_error("task_ctx lookup failed");
+
+	return taskc;
+}
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+	__type(key, u32);
+	__type(value, u64);
+	__uint(max_entries, P2DQ_NR_STATS);
+} stats SEC(".maps");
+
+static inline void stat_add(enum stat_idx idx, u64 amount)
+{
+	u32 idx_v = idx;
+
+	u64 *cnt_p = bpf_map_lookup_elem(&stats, &idx_v);
+	if (cnt_p)
+		(*cnt_p) += amount;
+}
+
+static inline void stat_inc(enum stat_idx idx)
+{
+	stat_add(idx, 1);
+}
+
+/*
+ * Returns if the task is interactive based on the last DSQ.
+ */
+static bool is_interactive(struct task_ctx *taskc)
+{
+	if (nr_dsqs_per_llc <= 1)
+		return false;
```

Comment:
Nothing is interactive in the single DSQ case? A comment here explaining why `false` is picked over `true` would be nice.